### PR TITLE
[PT-5] network

### PR DESCRIPTION
**Pull request by ibondarovich**
add authentication service for cheking user status
add sign in with google
add sign in with email and password
add sign up
add sign out
update navigation
add sign out to settings screen
add sign in and sign up screens

https://github.com/ibondarovich/FoodApp/assets/92872368/6257430d-7bd9-4880-a2e0-54d9f751f834

### CODE REVIEW

File: core_ui/lib/src/theme/app_dimens.dart [hunk 1]
```diff
@@ -25,19 +25,24 @@ class AppDimens {
...
 
   static const double iconSize16 = 16;
 
   static const double imageSize100 = 100;
 
   static const Offset boxShadowOffset = Offset(0, 20);
-}
\ No newline at end of file
+
+  static const List<double> textScales = [0.7, 1, 1.5];
```
Comment (Alexey-Lebedzeu) [inline core_ui/lib/src/theme/app_dimens.dart]: enum

File: data/lib/providers/local/local_provider.dart [hunk 2]
```diff
@@ -2,12 +2,17 @@ import 'package:data/entity/cart_item_entity/cart_item_entity.dart';
...
+  Future<void> saveScaleFactor(double input);
+  Future<void> saveUser(String input);
+  bool checkUser();
+  Future<void> removeUser();
+  List<CartItemEntity> getAllCartItems();
   List<DishEntity> getAllDishes();
-  Future<void> saveTheme(bool input);
+  void removeCartItem(int id);
+  double getScaleFactor();
   bool getTheme();
```
Comment (Alexey-Lebedzeu) [inline data/lib/providers/local/local_provider.dart]: saparate

File: data/lib/repositories/settings_repository_impl.dart [hunk 3]
```diff
@@ -17,4 +17,14 @@ class SettingsRepositoryImpl extends SettingsRepository{
   bool fetchTheme() {
     return _localProvider.getTheme();
   }
+  
+  @override
+  double fetchScaleFactor() {
+    return _localProvider.getScaleFactor();
```
Comment (Alexey-Lebedzeu) [inline data/lib/repositories/settings_repository_impl.dart]: check naming

File: domain/lib/repositories/user_repository.dart [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import 'package:domain/domain.dart';
+
+abstract class UserRepository {
+  Future<String> createUserWithEmailAndPassword(UserModel model);
+  Future<String> signInWithGoogle();
+  Future<void> signOut();
+  Future<String> signIn(UserModel model);
+  Future<void> saveUser(String input);
+  bool checkUser();
```
Comment (Alexey-Lebedzeu) [inline domain/lib/repositories/user_repository.dart]: naming

File: features/authentication_view/lib/src/ui/auth_screen.dart [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+class AuthScreen extends StatelessWidget{
+  const AuthScreen({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      backgroundColor: Theme.of(context).scaffoldBackgroundColor,
+      appBar: AppBar(
+        title: Image.asset(
+          'assets/img/logo.png',
```
Comment (Alexey-Lebedzeu) [inline features/authentication_view/lib/src/ui/auth_screen.dart]: ???

File: features/authentication_view/lib/src/ui/auth_screen.dart [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+          appRouter: appLocator.get<AppRouter>(), 
+          checkUserUseCase: appLocator.get<CheckUserUseCase>(),
+        ),
+        child: BlocConsumer<AuthenticationViewBloc, AuthenticationViewState>(
+          listener: (BuildContext context, AuthenticationViewState state) => {
+            if(state.authStatus.isNotEmpty){
+              ScaffoldMessenger.of(context)
+                ..hideCurrentSnackBar()
+                ..showSnackBar(
+                  SnackBar(
```
Comment (Alexey-Lebedzeu) [inline features/authentication_view/lib/src/ui/auth_screen.dart]: to core UI

**Review by Alexey-Lebedzeu — CHANGES_REQUESTED** [review]

File: core/lib/constants/string_constants.dart [hunk 7]
```diff
@@ -11,7 +11,7 @@ class StringConstants{
   static const String hiveBoxSettingsName = 'Settings';
   static const String hiveBoxScaleFactorName = 'ScaleFactor';
   static const String hiveBoxThemeName = 'Theme';
-
+  static const String hiveBoxUser = 'User';
```
Comment (dungeon-slave) [inline core/lib/constants/string_constants.dart]: naimng

File: core/lib/enums/text_scale_enum.dart [hunk 8]
```diff
@@ -0,0 +1,9 @@
+enum TextScaleType {
+  small(1),
+  medium(1.3),
+  large(1.5);
+
+  final double value;
+
+  const TextScaleType(this.value);
+}
```
Comment (dungeon-slave) [inline core/lib/enums/text_scale_enum.dart:9]: formatting

File: core/lib/validation/validation.dart [hunk 9]
```diff
@@ -0,0 +1,23 @@
+String? emailValidation(String email) {
+  if (email.isEmpty) {
+    return 'Please enter an email';
```
Comment (dungeon-slave) [inline core/lib/validation/validation.dart]: hardcode

File: core_ui/lib/src/widgets/app_image.dart [hunk 10]
```diff
@@ -2,37 +2,66 @@ import 'package:core_ui/core_ui.dart';
...
 import 'package:cached_network_image/cached_network_image.dart';
 
+enum ImageType {
+  png,
+  network,
+}
+
 class AppImage extends StatelessWidget {
-  final String imageURL;
+  final String filePath;
```
Comment (dungeon-slave) [inline core_ui/lib/src/widgets/app_image.dart:11]: make private

File: core_ui/lib/src/widgets/app_image_button.dart [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import 'package:flutter/material.dart';
+
+class AppImageButton extends StatelessWidget {
+  final Function()? onTap;
+  final String title;
```
Comment (dungeon-slave) [inline core_ui/lib/src/widgets/app_image_button.dart:5]: make private

File: core_ui/lib/src/widgets/app_snackbar.dart [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import 'package:core_ui/core_ui.dart';
+import 'package:flutter/material.dart';
+
+enum TitleTypeEnum { informational, error }
+
+void showAppSnackBar({
```
Comment (dungeon-slave) [inline core_ui/lib/src/widgets/app_snackbar.dart:6]: make widget

File: data/lib/providers/local/cart_data_provider_impl.dart [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      );
+    }
+    return await cartHiveBox.put(input.dishEntity.id, input);
+  }
+
+  @override
+  Future<void> updateCartItem(CartItemEntity input) async {
+    return await cartHiveBox.put(input.dishEntity.id, input);
+  }
+}
```
Comment (dungeon-slave) [inline data/lib/providers/local/cart_data_provider_impl.dart:38]: formatting

File: data/lib/providers/local/dish_data_provider.dart [hunk 14]
```diff
@@ -0,0 +1,6 @@
+import '../../entity/dish_entity/dish_entity.dart';
+
+abstract class DishDataProvider {
+  Future<void> saveAllDishes(List<DishEntity> input);
+  List<DishEntity> fetchAllDishes();
+}
```
Comment (dungeon-slave) [inline data/lib/providers/local/dish_data_provider.dart:6]: formatting

File: data/lib/providers/local/dish_data_provider_impl.dart [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+  final Box<DishEntity> menuHiveBox;
+
+  DishDataProviderImpl({
+    required this.menuHiveBox,
+  });
+
+  @override
+  Future<void> saveAllDishes(List<DishEntity> input) async {
+    await Future.wait(
+        input.map((item) async => await menuHiveBox.put(item.id, item)));
```
Comment (dungeon-slave) [inline data/lib/providers/local/dish_data_provider_impl.dart]: ,
type

File: data/lib/providers/local/theme_data_provider.dart [hunk 16]
```diff
@@ -0,0 +1,6 @@
+abstract class ThemeDataProvider{
+  Future<void> saveTheme(bool input);
+  Future<void> saveScaleFactor(double input);
+  double fetchScaleFactor();
+  bool fetchTheme();
+}
```
Comment (dungeon-slave) [inline data/lib/providers/local/theme_data_provider.dart:6]: formatting

File: data/lib/providers/local/theme_data_provider_impl.dart [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import 'package:core/core.dart';
+import 'package:data/providers/local/theme_data_provider.dart';
+import 'package:hive/hive.dart';
+
+class ThemeDataProviderImpl implements ThemeDataProvider{
+  final Box<bool> settingsHiveBox;
```
Comment (dungeon-slave) [inline data/lib/providers/local/theme_data_provider_impl.dart:6]: make private

File: data/lib/providers/local/user_data_provider.dart [hunk 18]
```diff
@@ -0,0 +1,5 @@
+abstract class UserDataProvider{
+  Future<void> saveUser(String input);
+  bool isUserExists();
+  Future<void> removeUser();
+}
```
Comment (dungeon-slave) [inline data/lib/providers/local/user_data_provider.dart:5]: formatting

File: data/lib/providers/local/user_data_provider_impl.dart [hunk 19]
```diff
@@ -0,0 +1,25 @@
+import 'package:data/providers/local/user_data_provider.dart';
+import 'package:hive/hive.dart';
+
+class UserDataProviderImpl implements UserDataProvider {
+  final Box<String> userHiveBox;
```
Comment (dungeon-slave) [inline data/lib/providers/local/user_data_provider_impl.dart:5]: make private

File: data/lib/repositories/cart_repository_impl.dart [hunk 20]
```diff
@@ -1,18 +1,18 @@
...
   CartRepositoryImpl({
-    required LocalProvider provider
+    required CartDataProvider provider
   }) : _provider = provider;
 
   @override
   List<CartItemModel> fetchAllCartItems() {
-    final List<CartItemEntity> response =  _provider.getAllCartItems();  
+    final List<CartItemEntity> response =  _provider.fetchAllCartItems();  
     return response.map((entity) => CartItemMapper.toModel(entity)).toList();
```
Comment (dungeon-slave) [inline data/lib/repositories/cart_repository_impl.dart]: type

File: features/authentication_view/lib/authentication_view.dart [hunk 21]
```diff
@@ -0,0 +1,3 @@
+library authentication_view;
+
+export 'src/ui/auth_screen.dart';
```
Comment (dungeon-slave) [inline features/authentication_view/lib/authentication_view.dart:3]: formatting

File: features/authentication_view/lib/src/bloc/bloc/bloc.dart [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+  ) {
+    emit(state.copyWith(isLoginScreenShowing: false));
+  }
+
+  void _onSignUp(
+    OnSignUpEvent event,
+    Emitter<AuthenticationViewState> emit,
+  ) async {
+    final bool result =
+        isValidated(state.emailValidationError, state.passwordValidationError);
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/bloc/bloc/bloc.dart]: ,

File: features/authentication_view/lib/src/ui/auth_screen.dart [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+  const AuthScreen({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      backgroundColor: Theme.of(context).scaffoldBackgroundColor,
+      appBar: AppBar(
+        title: const AppImage(
+          filePath: AppImages.logoImg,
+          scale: 2,
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/auth_screen.dart]: hardcode

File: features/authentication_view/lib/src/ui/components/app_text_field.dart [hunk 24]
```diff
@@ -0,0 +1,82 @@
+import 'package:core_ui/core_ui.dart';
+import 'package:flutter/material.dart';
+
+class AppTextField extends StatelessWidget {
+  final String? hintText;
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/app_text_field.dart:5]: make private

File: features/authentication_view/lib/src/ui/components/header.dart [hunk 25]
```diff
@@ -0,0 +1,46 @@
+import 'package:core_ui/core_ui.dart';
+import 'package:flutter/material.dart';
+
+class AuthHeader extends StatelessWidget {
+  final String mainTitle;
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/header.dart]: make private

File: features/authentication_view/lib/src/ui/components/login.dart [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+
+class LogInWidget extends StatefulWidget {
+  const LogInWidget({super.key});
+
+  @override
+  State<StatefulWidget> createState() => LogInState();
+}
+
+class LogInState extends State<LogInWidget> {
+  TextEditingController emailController = TextEditingController();
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/login.dart:17]: private

File: features/authentication_view/lib/src/ui/components/login.dart [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+                      AppTextField(
+                        hintText: StringConstants.enterEmail,
+                        controller: emailController,
+                        margin: const EdgeInsets.only(
+                          bottom: AppDimens.padding20,
+                        ),
+                        onChanged: (String value) => {
+                          BlocProvider.of<AuthenticationViewBloc>(context)
+                              .add(
+                            OnEmailChanged(value: value),
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/login.dart]: remove ,

File: features/authentication_view/lib/src/ui/components/login.dart [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+                        ),
+                      ),
+                      AppTextField(
+                        hintText: StringConstants.enterPassword,
+                        controller: passwordController,
+                        obscureText: true,
+                        onChanged: (String value) => {
+                          BlocProvider.of<AuthenticationViewBloc>(context)
+                              .add(
+                            OnPasswordChanged(value: value),
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/login.dart]: remove ,

File: features/authentication_view/lib/src/ui/components/signup.dart [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+
+class SignUpWidget extends StatefulWidget {
+  const SignUpWidget({super.key});
+
+  @override
+  State<StatefulWidget> createState() => SignUpState();
+}
+
+class SignUpState extends State<SignUpWidget> {
+  TextEditingController emailController = TextEditingController();
```
Comment (dungeon-slave) [inline features/authentication_view/lib/src/ui/components/signup.dart:17]: private

File: features/main_view/lib/main_view.dart [hunk 30]
```diff
@@ -0,0 +1,3 @@
+library main_view;
+
+export 'src/ui/main_view_screen.dart';
```
Comment (dungeon-slave) [inline features/main_view/lib/main_view.dart:3]: formatting

File: navigation/lib/src/app_router/app_router.dart [hunk 31]
```diff
@@ -16,44 +19,59 @@ part 'app_router.gr.dart';
...
+      ],
     ),
   ],
 )
 
-class AppRouter extends _$AppRouter {}
\ No newline at end of file
+class AppRouter extends _$AppRouter {
+  AppRouter({required super.authGuard});
+}
```
Comment (dungeon-slave) [inline navigation/lib/src/app_router/app_router.dart:77]: formatting

File: features/authentication_view/lib/src/ui/components/login.dart (hunk 26 above)
Comment (Alexey-Lebedzeu) [inline features/authentication_view/lib/src/ui/components/login.dart]: final

File: features/authentication_view/lib/src/ui/components/signup.dart (hunk 29 above)
Comment (Alexey-Lebedzeu) [inline features/authentication_view/lib/src/ui/components/signup.dart]: final

**Pull request merged**